### PR TITLE
Help modders deal with object invalidation

### PR DESCRIPTION
**Pull request by appgurueu**
Implements part of what I have outlined in #14687 (related: #13297):

* "Fixes" raycasts: Invalid objects should be skipped.
* Adds a method `is_valid` to idiomatically check for object validity.
* Adds iterator variants of `get_objects_(inside_radius|in_area)`, which skip invalid objects (and are also more convenient to use).

### CODE REVIEW

**Comment by sfan5** [comment]
Adding a method does not qualify as bugfix, but the first two commits would be useful for 5.9 so how about splitting it?

> * Adds iterator variants of `get_objects_(inside_radius|in_area)`, which skip invalid objects

But the function already skips invalid objects, so this is only convenience, right?

File: doc/lua_api.md [hunk 1]
```diff
@@ -7815,6 +7821,10 @@ child will follow movement and rotation of that bone.
 
 ### Methods
 
+* `is_valid()`: returns whether the object is valid.
+   Objects may be invalidated either through explicit removal,
+   or implicitly between server steps by the engine.
```
Comment (sfan5) [inline doc/lua_api.md:7826]: IIRC there's an "advice on object invalidation" section that should be updated alongside

it's also not mentioned that any objectrefs returned by engine functions are guaranteed to be valid at the time of returning
which then means that in most cases you won't need to deal with this

**Comment by rubenwardy** [comment]
> But the function already skips invalid objects, so this is only convenience, right?

The current function returns a table of objects. Some of these objects may be invalidated between building the table and reaching the object in the for loop

File: doc/lua_api.md [hunk 2]
```diff
@@ -7815,6 +7821,10 @@ child will follow movement and rotation of that bone.
...
+   **The below methods should only be called for valid objects.**
```
Comment (sfan5) [inline doc/lua_api.md:7827]: also this explicitly makes any method calls on invalid objects UB. that includes many existing mods
do we want to do that?

IMO at the very least generic getters (e.g. `get_pos`) should be allowed to be called on invalid references

**Comment by sfan5** [comment]
> The current function returns a table of objects. Some of these objects may be invalidated between building the table and reaching the object in the for loop

Ah, that's easy to miss. The docs should explain the benefit of the new functions more clearly.

**Comment by appgurueu** [comment]
I agree that it may be debatable whether the new iterator variants (and `:is_valid`) are a "bug fix", but they are certainly helpful in eliminating bugs in mod code. (I also think that the implementation is simple enough that there is little risk in it.)

I can split if you want me to, though.

**Comment by sfan5** [comment]
After finding out that it's not just a simple helper but rather fixes an easy to miss edge case I'm okay with keeping it as a bug fix.

File: doc/lua_api.md (hunk 2 above)
Comment (appgurueu) [reply doc/lua_api.md:7827]: I've checked the code and all getters consistently return nothing when called on invalid objects. It's probably fine to keep that for now, I wouldn't want to deprecate it yet - too much work for everyone involved.

(The "should" was intended as in "this is best practice if you're writing a mod for a recent Minetest version today", not as in "this is deprecated and will be broken soon". I have clarified.)

File: doc/lua_api.md (hunk 1 above)
Comment (appgurueu) [reply doc/lua_api.md]: thanks for pointing that out, edited the section

> it's also not mentioned that any objectrefs returned by engine functions are guaranteed to be valid at the time of returning

this seems to be already (somewhat implicitly) mentioned in said section so we're good?

Comment (sfan5) [reply doc/lua_api.md]: > this seems to be already (somewhat implicitly) mentioned in said section so we're good?

sure.
can you add a reference to that section too?

**Comment by Zughy** [comment]
Why not sinply overwrite the current get functions? Like, I can't think of any useful case where mods want to iterate invalid objects as well. On the contrary, I always have to put checks in my for loops in order to avoid crashes

**Comment by appgurueu** [comment]
> Why not sinply overwrite the current get functions?

Their API unfortunately makes this infeasible: They return a "list" of object refs.

Now the edge case happens when you do something like this, for example:

```lua
for _, obj in ipairs(minetest.get_objects_inside_radius(...)) do
    obj:punch(...)
end
```

Because punching an object can invalidate other objects in the list (such as children) via side effects. The problem is that these objects are already in the table, so there's no (proper) way to get rid of them after the fact. They will be visited in the traversal.

With an iterator, this becomes cleaner, and the iterator can now check for and skip invalid objects. But this obviously requires modders to use the iterator rather than the table variant, hence is a "breaking" change and requires the addition of a new function which does the right thing.

File: doc/lua_api.md [hunk 3]
```diff
@@ -6123,12 +6123,23 @@ Environment access
...
-    * returns a list of ObjectRefs.
+* `minetest.get_objects_inside_radius(center, radius)`
+    * returns a list of ObjectRefs
     * `radius`: using a Euclidean metric
-* `minetest.get_objects_in_area(pos1, pos2)`
-    * returns a list of ObjectRefs.
-    * `pos1` and `pos2` are the min and max positions of the area to search.
+    * **Warning**: If you modify objects (f.e. punch them),
+      this may cause later objects in the list (f.e. children) to become invalid.
+      Use `minetest.objects_inside_radius` instead to iterate only valid objects
```
Comment (sfan5) [inline doc/lua_api.md:6131]: I would reword this
```suggestion
    * **Warning**: Any kind of interaction with the environment or other APIs
      can cause later objects in the list to become invalid while you're iterating it.
      (e.g. punching an entity removes its children)
      It is recommend to use `minetest.objects_inside_radius` instead, which
      transparently takes care of this possibility.
```

File: games/devtest/mods/unittests/entity.lua [hunk 4]
```diff
@@ -71,13 +71,13 @@ local function test_entity_lifecycle(_, pos)
 
 	-- with binary in staticdata
 	local obj = core.add_entity(pos, "unittests:callbacks", "abc\000def")
+	assert(obj:is_valid())
```
Comment (sfan5) [inline games/devtest/mods/unittests/entity.lua:74]:
```suggestion
	assert(obj and obj:is_valid())
```

File: games/devtest/mods/unittests/entity.lua [hunk 5]
```diff
@@ -130,3 +130,49 @@ local function test_entity_attach(player, pos)
...
+
+local function test_entity_raycast(_, pos)
+	local obj1 = core.add_entity(pos, "unittests:dummy")
+	local obj2 = core.add_entity(pos:offset(1, 0, 0), "unittests:dummy")
+	local raycast = core.raycast(pos:offset(-1, 0, 0), pos:offset(2, 0, 0), true, false)
+	for pt in raycast do
+		if pt.type == "object" then
+			assert(pt.ref == obj1)
+			obj2:remove()
+			obj1 = nil -- object should be hit exactly one
```
Comment (sfan5) [inline games/devtest/mods/unittests/entity.lua:151]: who removes obj1 after this test?

File: games/devtest/mods/unittests/entity.lua [hunk 6]
```diff
@@ -130,3 +130,49 @@ local function test_entity_attach(player, pos)
...
+	local obj2 = core.add_entity(pos, "unittests:dummy")
+	-- As soon as we find one of the objects, we invalidate the other.
+	for obj in iterator do
+		assert(obj:is_valid())
+		if obj == obj1 then
+			obj2:remove()
+		elseif obj == obj2 then
+			obj1:remove()
+		end
+	end
```
Comment (sfan5) [inline games/devtest/mods/unittests/entity.lua:169]: test won't fail if iterator is unexpectedly empty

**Review by sfan5 — APPROVED** [review]
Fine by me.

**Comment by Zughy** [comment]
Wouldn't this fix the ghost mode issue completely?

**Comment by appgurueu** [comment]
> Wouldn't this fix the ghost entities issue completely?

Ultimately, I don't know. It fixes ghost entity issues related to this mistake, if the API is used properly.

I would say it would probably make the issue a "possible close" under suspicion that this is the (main?) root cause for the issue.

I can't rule out the possibility that there are other places where "ghost entities" may appear, though, i.e. that there is an underlying bug not fixed by this.
